### PR TITLE
feat(ec2): add im4gn and is4gen instances

### PR DESCRIPTION
**Pull request by robertd**
Newly announced `im4gn` and `is4gen` storage optimized instances:
https://aws.amazon.com/blogs/aws/new-storage-optimized-amazon-ec2-instances-im4gn-and-is4gen-powered-by-aws-graviton2-processors/

image

image

CFN docs have already been updated:
https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-ec2-instance.html#cfn-ec2-instance-instancetype

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

**Comment by robertd** [comment]
Build currently failing. Probably related to: https://github.com/aws/jsii/issues/3208

```
error JSII8001: Enum members must be ALL_CAPS. Rename "@aws-cdk/aws-ec2.InstanceClass.IM4GN" to "IM4_G_N"
error JSII8001: Enum members must be ALL_CAPS. Rename "@aws-cdk/aws-ec2.InstanceClass.IS4GEN" to "IS4_GE_N"
```

**Comment by robertd** [comment]
https://github.com/aws/jsii/pull/3209 landed... so this will probably go away once `jsii` gets bumped to `1.47.0` whenever that one gets released.

File: packages/@aws-cdk/aws-ec2/lib/instance-types.ts [hunk 1]
```diff
@@ -362,6 +362,16 @@ export enum InstanceClass {
...
 
+  /**
+   * Storage optimized instances powered by Graviton2 processor, 4th generation
+   */
+  IM4GN = 'im4gn',
+
+  /**
+   * Storage optimized instances powered by Graviton2 processor, 4th generation
+   */
+  IS4GEN = 'is4gen',
```
Comment (jumic) [inline packages/@aws-cdk/aws-ec2/lib/instance-types.ts:373]: Could you please add symbolic names for both enums, too? Thanks.

Comment (robertd) [reply packages/@aws-cdk/aws-ec2/lib/instance-types.ts:373]: Can you suggest one of IS4GEN?

Comment (jumic) [reply packages/@aws-cdk/aws-ec2/lib/instance-types.ts:373]: Based on the [product page](https://aws.amazon.com/de/ec2/instance-types/i4g/), maybe we could use `STORAGE4_HIGH_PERFORMANCE =  'im4gn',` and `STORAGE4_LOW_COST =  'is4gen',`?

Unfortunately, I can't find any further information on how the abbreviations are composed. We can also wait for further feedback or ideas.

**Comment by robertd** [comment]
Depends on #17904

File: packages/@aws-cdk/aws-ec2/lib/instance-types.ts (hunk 1 above)
Comment (robertd) [reply packages/@aws-cdk/aws-ec2/lib/instance-types.ts:383]: @njlynch wdyt?

Comment (njlynch) [reply packages/@aws-cdk/aws-ec2/lib/instance-types.ts:383]: https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/instance-types.html#instance-hardware-specs

> Instance type names combine the instance family, generation, and size. They can also indicate additional capabilities, such as:
> * a – AMD processors
> * g – AWS Graviton processors
> * i – Intel processors
> * d – Instance store volumes
> * n – Network optimization
> * b – Block storage optimization
> * e – Extra storage or memory
> * z – High frequency

So these instances are storage-optimized, Graviton-driven, and network-optimized, with the latter also having extra storage/memory. That's a mouthful to stuff into an enum name.

`STORAGE4_GRAVITON_NETWORK_OPTIMIZED` and ``STORAGE4_GRAVITON_NETWORK_STORAGE_OPTIMIZED` would be some suggestions, but I don't love them. Open to counter proposals.

Comment (robertd) [reply packages/@aws-cdk/aws-ec2/lib/instance-types.ts]: Sounds good to me @njlynch. PR updated.

**Review by njlynch — APPROVED** [review]

**Pull request merged**